sanitize html in applab [#104341490]

davidsbailey opened the pull request:
pivotal item: https://www.pivotaltracker.com/projects/1256654/stories/104341490
dev doc: https://docs.google.com/document/d/1JuIJT0ckZqxtR07_u7XpjnoB8Xy8NR6BCnOhcKL7D4Y/edit

@Bjvanminnen could you please take a look at this PR and the dev doc above, and let me know if you agree with the approach? 

FYI @wjordan @philbogle @cpirich @islemaster please feel free to chime in!

Bjvanminnen commented:
What are peoples' thoughts on trying to clean up the html vs. just rejecting apps where we find "bad" html. The latter approach is much more heavy handed, but I thought was the one we were leaning towards last time we'd talked about this.

I guess ideal approach would be that if someone unintentionally created a "bad" app (which is ideally not too common), we give them an explanation of what's "bad". Not sure how easy this is/isnt to do with this library (or others). Thoughts?

Bjvanminnen commented on apps/src/utils.js:
```diff
@@ -1,6 +1,7 @@
+/* global $, define */
 
 var xml = require('./xml');
+var sanitizeHtml = require('sanitize-html');
```
You'll need to also make sure this ends up in package.json (`npm install --save-dev sanitize-html` is the easiest way to accomplish that).

Bjvanminnen commented:
Another thought I just had...I wonder if we should log to new relic when we end up sanitizing away some html (or in the world where we just block "bad" html). This way we can have some insight into how often we hit this.

davidsbailey commented:
Generally, the tools I have found will sanitize the HTML rather than accepting/rejecting it. Based on feedback, I'm going to explore how reliably we can use the sanitizer to detect when parts of HTML are rejected, and how precisely we can give useful warnings about the differences. That should inform answers to the other questions that have been raised.

davidsbailey commented:
@Bjvanminnen it took a bit of hackery to reliably detect non-cosmetic changes when sanitizing html: https://github.com/code-dot-org/code-dot-org/commit/feaef2e6b5c631b6d8fa9e8cca9b2aa17c8b04b7 as implemented this could be used to notify new relic if initial html ever contained invalid html. 

Please let me know if you think this looks too hacky to proceed. Alternatives include:
1. try to make the 1-line fix to properly handle `allowedSchemes: false` in the upstream github repository: https://github.com/punkave/sanitize-html/blob/master/index.js#L295
2. incorporate the source code into our repository and make the 1-line change locally
3. use a different library

davidsbailey commented:
<img width="1737" alt="screen shot 2016-01-20 at 12 38 33 pm" src="https://cloud.githubusercontent.com/assets/8001765/12462724/b1a62ca6-bf74-11e5-9ce4-887ffa1d0b2b.png">

Bjvanminnen commented on apps/src/utils.js:
```diff
@@ -68,6 +69,92 @@ exports.escapeHtml = function(unsafe) {
+    return 0;
+  };
+
+  var processed = sanitizeHtml(unsafe, {
```
if safe and unsafe are already the same, can we skip this extra processing step?

Bjvanminnen commented:
Regarding hackery, it's probably worth at least opening an issue against the upstream repo, and then including a PR if you are so inclined :)

Bjvanminnen commented on apps/src/utils.js:
```diff
@@ -68,6 +69,92 @@ exports.escapeHtml = function(unsafe) {
+  return safe;
+};
+
+/**
```
I might put these sanitizeHtml related methods in their own file instead of lumping them with the rest of utils.

Bjvanminnen commented:
I'd leave the hackery in for now, perhaps with a comment linking to the issue you open in their repo, so that we could theoretically remove the hackery if/when fixed. Otherwise, lgtm

davidsbailey commented:
addressed feedback and added tests. tests have more warning spew unfortunately:

```
  Level tests
    ../solutions/applab/ec_uicontrols.js
      ✓ UI controls (578ms)
      ✓ getText and setText on text labels. (221ms)
      ✓ button (204ms)
      ✓ setPosition (203ms)
WARNING: Line: 1: button() id parameter refers to an id (button1) which already exists.
      ✓ testDuplicateNamesWarning (185ms)
ERROR: Line: 1: Error: button() id parameter refers to an id (runButton) which is already in use outside of Applab. Choose a different id.
      ✓ testIllegalNameError (207ms)
      ✓ setSize (190ms)
      ✓ cosmetic changes in html sanitization do not cause warnings (176ms)
WARNING: Line: 1: The following lines of HTML were modified or removed:
<img id="image6" src="javascript:alert()" />
<script>alert()
</script>
original html:
<div id="container"><img id="image3" src=""><img id="image4" src="//localhost:8001/apps/static/flappy_promo.png" /><img id="image5" src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" /><img id="image6" src="javascript:alert()" /><script>alert()</script></div>
modified html:
<div id="container"><img id="image3" src /><img id="image4" src="//localhost:8001/apps/static/flappy_promo.png" /><img id="image5" src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw==" /><img id="image6" /></div>
      ✓ substantial changes in html sanitization do cause warnings (209ms)
```

Bjvanminnen commented:
Seems like warning spew should be reasonably easy to address, though I have vague recollections of attempting to do so and failing. In any case, it's unfortunate but not pressing.

Bjvanminnen commented on apps/src/applab/sanitizeHtml.js:
```diff
@@ -0,0 +1,87 @@
+ * @param {function(removed, unsafe, safe)} warn Optional function to call if
+ *     any unsafe html was removed from the output.
+ */
+module.exports = function(unsafe, warn) {
```
I would still name this function for clarity, i.e.

```
module.exports = function sanitizeHtml(unsafe, warn)
```

Bjvanminnen commented on apps/src/applab/designMode.js:
```diff
@@ -473,7 +474,8 @@ designMode.parseFromLevelHtml = function(rootEl, allowDragging, prefix) {
     return;
   }
-  var levelDom = $.parseHTML(Applab.levelHtml);
+  var sanitized = sanitizeHtml(Applab.levelHtml);
```
Do we not get any feedback if Applab.levelHtml sanitizes away some contents? I think we'd discussed perhaps just failing in that case? Seems reasonable if the plan is just to do that in a future commit so that these changes are ready for tomorrow.

davidsbailey replied:
I agree that there is something aesthetically pleasing about rejecting rather than sanitizing level html written by a hacker, but is there any other advantage of failing in that case? Both provide the same level of security against hackers. Neither has an advantage for legitimate users when the app is working properly. So the only difference is that if there is a programming error, we would block all `levelHtml` for legitimate users rather than allowing the legitimate part of it through.

davidsbailey replied:
actually I think this is the case in which we had talked about opening a newrelic or honeybadger error. Here is a bug to track that: https://www.pivotaltracker.com/story/show/112187121

Bjvanminnen replied:
Yeah, mostly I just want to make sure that there's some way we discover these errors. Logging to the console we may or may not notice (it might be sufficient), silently sanitizing we would not notice :) Logging a bug to address this in a future commit in some way seems like the right approach :+1:

Bjvanminnen commented on apps/test/solutions/applab/ec_uicontrols.js:
```diff
@@ -189,5 +193,98 @@ module.exports = {
+        result: true,
+        testResult: TestResults.FREE_PLAY
+      },
+    }
```
One thing we could if we wanted to reduce spam here would be to create a test outside of levelTests that tests just santizeHtml.js. It would then be more of a true unit test, feeding inputs to sanitizeHtml, and validating that it gets the expected output.

We could then also have one or more simple level tests that are more of integration tests, validating that things do actually end up in our console.

In any case, none of these need block checkin.

davidsbailey replied:
proposal: disable console.log on applab debug console errors in unit tests. https://www.pivotaltracker.com/n/projects/1256654

Bjvanminnen commented:
Few comments, but generally lgtm assuming circle ci failures are okay

davidsbailey commented:
CircleCI failures indicated a mochaTest error, but mochaTests passed for me when I ran them locally. I'll keep an eye on staging when I merge.
